### ISSUE TITLE
-race does not obey go:nocheckptr

### ISSUE THREAD

**Issue by chewxy** [golang/go#42880]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.15.5 linux/amd64
</pre>

### Does this issue reproduce with the latest release?

yes (even with `gotip`)

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE="off"
GOARCH="amd64"
GOBIN=""
GOCACHE="PATH/TO/go-build"
GOENV="PATH/TO/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="PATH/TO/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="GOPATH"
GOPRIVATE="PRIVATE PROXY HERE"
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build533267197=/tmp/go-build -gno-record-gcc-switches"

</pre></details>

### What did you do?

I ran a test to test racy behaviours - This function had a checkptr error:

https://github.com/gorgonia/tensor/blob/9166554e0f40794485c199a8ecd4d692d24dbe73/consopt.go#L136-L171

The test is this: https://github.com/gorgonia/tensor/blob/9166554e0f40794485c199a8ecd4d692d24dbe73/consopt_test.go#L19-L33

Most notably, the function had a `//go:nocheckptr` directive attached to it. It's been there for the purpose of CUDA - since CUDA memory is not accessible by the Go runtime (technically it can, but it has to be run on the same thread as the CUDA runtime), I omitted the `checkptr` tests because they are likely to fail.

In the past I had manually used `-d=checkptr`. That seems to obey the `go:nocheckptr` directive. 

### What did you expect to see?
No checkptr errors. 

### What did you see instead?

Checkptr errors.

**Comment by ianlancetaylor** [comment]
Please show exactly what you saw.  Thanks.

Please also describe exactly how to reproduce the problem.  Thanks.

**Comment by chewxy** [comment]
Updated with what exactly I did, and what I saw

**Comment by ianlancetaylor** [comment]
CC @mdempsky

**Comment by chewxy** [comment]
A bit more to add. If I understand the Go semantics correctly, the test, as written shouldn't fail the checkptr either. But I am willing to concede that there are parts which I do not know as well

**Comment by mdempsky** [comment]
It's not safe to convert a `uintptr`-typed variable to `unsafe.Pointer`: https://github.com/gorgonia/tensor/blob/9166554e0f40794485c199a8ecd4d692d24dbe73/consopt.go#L156. cmd/compile is correct to emit an error about that code.

As for why `//go:nocheckptr` didn't suppress the error, it's because the error is actually within a function literal, and `//go:` directives only apply to declared functions. Perhaps we should revisit that though.

**Comment by chewxy** [comment]
> It's not safe to convert a uintptr-typed variable to unsafe.Pointer: https://github.com/gorgonia/tensor/blob/9166554e0f40794485c199a8ecd4d692d24dbe73/consopt.go#L156. cmd/compile is correct to emit an error about that code.

Yeah, this part I am aware of. The only time this is used is for CUDA code, really, which due to some complications, need `unsafe.Pointer` (until I finish integrating CUDA11... but that's gonna take time)

> As for why //go:nocheckptr didn't suppress the error, it's because the error is actually within a function literal, and //go: directives only apply to declared functions. Perhaps we should revisit that though.

Can I put `//go:nocheckptr` on top of `f := func(t Tensor) {` ?

**Comment by cuonglm** [comment]
> Can I put //go:nocheckptr on top of f := func(t Tensor) { ?

You can't.

**Comment by mdempsky** [comment]
> Yeah, this part I am aware of.

In your last message you wrote "A bit more to add. If I understand the Go semantics correctly, the test, as written shouldn't fail the checkptr either." I was explaining why the checkptr error is correct. Your earlier understanding was incorrect.

> Can I put //go:nocheckptr on top of f := func(t Tensor) { ?

No. At the moment, `//go:nocheckptr` only applies to declared functions.

You could use a helper function like:

```
//go:nocheckptr
func asUnsafePointer(x uintptr) unsafe.Pointer { return unsafe.Pointer(x) }
```

But again, I'll emphasize your test case is in fact unsafe.
